addons/mail/migrations/12.0.1.0/end-migration.py

```diff
@@ -35,29 +36,26 @@ def fill_mail_thread_message_main_attachment_id(env):
     thread_models = [k for k in env.registry if issubclass(
         type(env[k]), type(env['mail.thread'])) and env[k]._auto]
     for model in thread_models:
-        records = env[model].with_context(active_test=False).search(
-            [('message_main_attachment_id', '=', False)])
-        attachment_obj = env['ir.attachment'].with_context(
-            prefetch_fields=False,
+        openupgrade.logged_query(
+            env.cr, """
+            UPDATE %s t
+            SET message_main_attachment_id = ia.id
+            FROM (
+                SELECT id, res_id
```

Comment:
Why do you need 3 nested select? For sure you can reduce this expression.

Reply:
Well, if you know how to reduce it, then propose it. By the way, this query is already superfast, I don't see the need to break my brain to find a less nested query.

Reply:
It's more a question of being more readable than performance. It's hard to deduce what is this doing...

Reply:
I will not break my brain to make it more readable. What matters to me is that the query really does exactly what the previous code did but with efficacy.

But if you want an explanation: the query groups the attachments by their kind of mimetype and res_id. For each res_id, you will have one, two or three of this groups if they exist (and labeled them by 1, 2 or 3), but the query always will select the existing group with the lower label. The `max(id)` is because the order in `ir.attachment ` is defined as `id DESC`.

Reply:
OK, thanks for the explanation. Just put as code comment and it's enough for me.

Reply:
done